torch/distributed/fsdp/_flat_param.py

```diff
@@ -0,0 +1,474 @@
+            module parameter; for any parameter ``p`` that is not sharded
+            across ranks, this will be [0, ``p.numel()``-1].
+        _shard_indices (Tuple[int, int]): [start, end] indices (in units of
+            parameters) for this rank's shard of the original model parameters,
+            where the parameters follow the order in which they were originally
+            flattened; this indexes appropriately into any data structure that
+            follows the flattening order (e.g. ``_param_infos``, ``_numels``,
+            etc.).
+        _shard_numel_padded (int): Numel padded for this rank's sharded
+            flattened parameter.
+
+        _local_shard (Tensor): Sharded flattened parameter with padding.
+        _full_param_padded (Tensor): Unsharded flattened parameter with
+            padding.
+        _shard_bwd_hook (Tuple[AccumulateGrad, RemovableHandle]): Flattened
+            parameter's :class:`AccumulateGrad` object and post-backward hook
+            handle.
+        _mp_shard (Tensor): Reduced-precision flattened parameter with padding.
+        _cpu_grad (Tensor): Sharded gradient with padding stored on CPU.
+        _saved_grad_shard (Tensor): Sharded gradient with padding from previous
+            iterations for gradient accumulation without :meth:`no_sync`.
+    """
+
+    def init_metadata(
+        self,
+        param_infos: List[ParamInfo],
+        numels: List[int],
+        shapes: List[torch.Size],
+        prefixed_param_names: List[str],
+        shared_param_infos: List[SharedParamInfo],
+    ) -> None:
+        """
+        Initializes attributes holding metadata about the original parameters
+        comprising the flattened parameter.
+
+        We expose this method separate from the constructor to keep the
+        constructor only responsible for the flattened parameter's tensor data.
+        This method should only be called once per model, while the constructor
+        may be called multiple times, e.g. when reloading from a checkpoint, in
+        which case only the tensor data needs to be passed to the constructor.
+        Since :meth:`load_state_dict` is implemented via :meth:`copy_`, the
+        metadata is correctly assumed to be unchanged.
+
+        Args:
+            See the Attributes in the class docstring.
+        """
+        assert len(param_infos) == len(numels)
+        assert len(param_infos) == len(shapes)
+        assert len(param_infos) == len(prefixed_param_names)
+        self._num_params = len(param_infos)
+        self._param_infos = tuple(param_infos)
+        self._numels = tuple(numels)
+        self._shapes = tuple(shapes)
+        self._prefixed_param_names = tuple(prefixed_param_names)
+        self._shared_param_infos = tuple(shared_param_infos)
+        self._is_sharded = False
+        self._unsharded_size = self.size()
+
+    @property
+    def _flat_param_name(self) -> str:
```

Comment:
This is a good point. I used this `_flat_param_name` frequently when debugging, but it may not be tractable for real usage because it will be too long. For this PR, the name is not actually used anywhere, so I can remove it. If needed for the follow-up PR, we can reintroduce it.